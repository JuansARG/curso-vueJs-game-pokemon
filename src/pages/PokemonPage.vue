<template>
    
    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <h1>Quien es este pokemon?</h1>
    
        <CompPokemonPicture 
            :pokemon-id="pokemon.id"
            :show-pokemon="showPokemon"/>

        <CompPokemonOptions
            v-if='showOptions'
            :pokemons="pokemonArr"
            @selection="checkAnswer"/>

        <template v-if="showAnswer">
            <h2>{{ message }}</h2>
            <button @click="newGame">
                Nuevo Juego
            </button>
        </template>

    </div>
    
</template>

<script>

import CompPokemonPicture from '@/components/CompPokemonPicture'
import CompPokemonOptions from '@/components/CompPokemonOptions'
import getPokemonsOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        CompPokemonPicture,
        CompPokemonOptions,
    },
    data(){
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
            showOptions: true
        }
    },
    methods: {
        async mixPokemonArray(){
            this.pokemonArr = await getPokemonsOptions()

            const rndInt = Math.floor( Math.random() * 4 )
            this.pokemon = this.pokemonArr[rndInt]

        },
        checkAnswer( selectedId ){
            this.showPokemon = true
            this.showAnswer = true
            this.showOptions = false

            if( selectedId === this.pokemon.id ){
                this.message = `Correcto, es ${ this.pokemon.name.toUpperCase() }`
            }else{
                this.message = `Oops, era ${ this.pokemon.name.toUpperCase() }`
            }
            
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.showOptions = true
            this.mixPokemonArray()
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>

<style>

</style>