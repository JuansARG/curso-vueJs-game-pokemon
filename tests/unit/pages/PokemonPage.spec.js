import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from '@/pages/PokemonPage'
import { mocksPokemons } from "../mocks/pokemons.mock"


describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
        jest.clearAllMocks()
    })

    test('Debe hacer match con el snapshot antes de cargar los pokemons', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('Debe hacer match con el snapshot cuando cargan los pokemons', () => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: mocksPokemons,
                    pokemon: mocksPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    showOptions: true
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()

    })

    test('Debe llamar mixPokemonArray al montar', () => {

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        const wrapper = shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled()

    })

    test('Debe mostrar los componentes CompPokemonPicture y CompPokemonOptions', () => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: mocksPokemons,
                    pokemon: mocksPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    showOptions: true
                }
            }
        })

        const compPokemonPicture = wrapper.findComponent('comp-pokemon-picture-stub')
        const compPokemonOptions = wrapper.findComponent('comp-pokemon-options-stub')

        expect( compPokemonPicture.exists() ).toBeTruthy()
        expect( compPokemonPicture.attributes('pokemonid') ).toBe( String(mocksPokemons[0].id) )

        expect( compPokemonOptions.exists() ).toBeTruthy()
        expect( compPokemonOptions.attributes('pokemons') ).toBeTruthy()
        

    })

    test('Pruebas con checkAnswer en caso de respuesta correcta', async() => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: mocksPokemons,
                    pokemon: mocksPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    showOptions: true
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBeTruthy()
        expect( wrapper.find('h2').text() ).toBe( `Correcto, es ${ mocksPokemons[0].name.toUpperCase() }` )

    })

    test('Pruebas con checkAnswer en caso de respuesta incorrecta', async() => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: mocksPokemons,
                    pokemon: mocksPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    showOptions: true
                }
            }
        })

        await wrapper.vm.checkAnswer(4)

        expect( wrapper.vm.message ).toBe( `Oops, era ${ mocksPokemons[0].name.toUpperCase() }` )
    })

})
