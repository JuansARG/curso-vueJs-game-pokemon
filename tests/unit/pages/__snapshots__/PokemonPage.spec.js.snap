// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe hacer match con el snapshot antes de cargar los pokemons 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage Component Debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>Quien es este pokemon?</h1>
  <comp-pokemon-picture-stub pokemonid="1" showpokemon="false"></comp-pokemon-picture-stub>
  <comp-pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></comp-pokemon-options-stub>
  <!--v-if-->
</div>
`;
